Support for processing Elastic Beanstalk worker SQS Daemon requests in Active Job

MrWellington opened the pull request:
For [feature-request 54](https://github.com/aws/aws-sdk-rails/issues/54):

Here's a first pass at an implementation for this feature. In short the intent is for the middleware to be added in the presence of an "AWS_PROCESS_BEANSTALK_WORKER_REQUESTS" environment variable (can change if that violates any conventions). This middleware should be able to retrieve Active Job SQS messages forwarded to the application by the daemon and pass them to their associated jobs. There is also code to handle scheduled jobs created by the worker environment itself. More information can be found in the readme updates.

Note that the [Active Elastic Job](https://github.com/tawan/active-elastic-job) gem that influenced this design involved an additional security measure wherein the queuing backend set a message digest -- and the middleware verified it -- using the same Rails SECRET_KEY_BASE value in both environments. I didn't consider this to be essential (thought restricting senders to localhost/docker host was more than adequate), and to reduce impact to the existing code did not implement it. Wanted to get it on your radar though - if you agree it is not essential at present might be a nice additional security layer to add later.

mullermp commented:
Thanks for opening a PR! We will review this shortly.

alextwoods commented:
Thanks for opening!  I'll take a look today.

I do agree with the decision not to implement the additional check against the Rails secret key - I think the restricted sending permissions should be sufficent, but it is an interesting idea and something we may want to add at some point (though as you point out, it wouldn't be backwards compatible and would need to be added as an opt-in/optional configuration).

alextwoods commented on lib/aws/rails/middleware/elastic_beanstalk_worker_listener.rb:
```diff
@@ -0,0 +1,90 @@
+
+      def execute_job(request)
+        # Jobs queued from the Active Job SQS adapter contain the JSON message in the request body.
+        job = JSON.parse(request.body.string)
```
I'd recommend using the [JobRunner](https://github.com/aws/aws-sdk-rails/blob/master/lib/aws/rails/sqs_active_job/job_runner.rb) here - it does functionally do the same thing, but lets us keep that logic abstracted.  It also uses the Aws::Json (which lets users use whichever json engine they have configured rather).

MrWellington replied:
Ah yeah makes sense - completely missed that class while working on this! Will update.

alextwoods replied:
Agree with your comments lower down on not using the JobRunner here since it takes the sqs message -  your update here looks good.

alextwoods commented on lib/aws/rails/railtie.rb:
```diff
@@ -64,5 +68,21 @@ def self.instrument_sdk_operations
+    # This will only be added in the presence of the AWS_PROCESS_BEANSTALK_WORKER_REQUESTS environment variable.
+    # The expectation is this variable should only be set on EB worker environments.
+    def self.add_sqsd_listener_middleware(app)
+      is_eb_worker_hosted = ENV['AWS_PROCESS_BEANSTALK_WORKER_REQUESTS']
```
You can use the `Aws::Util.str_2_bool` method for this.

MrWellington replied:
Sounds good - will do.

alextwoods commented on lib/aws/rails/middleware/elastic_beanstalk_worker_listener.rb:
```diff
@@ -0,0 +1,90 @@
+module Aws
+  module Rails
+    # Middleware to handle requests from the SQS Daemon present on Elastic Beanstalk worker environments.
+    class ElasticBeanstalkWorkerListener
```
I'd argue this isn't really a Listener -  maybe `EbsSqsActiveJobMiddleware` or something like that - trying to keep the name from being too long while being descriptive.  This more specific than handling SQS Daemon requests - it only handles requests that are translatable to ActiveJob (cron or sqs).

MrWellington replied:
Yeah I didn't care for the name and was waffling on it...I like your proposed one better.

alextwoods commented on test/aws/rails/middleware/elastic_beanstalk_worker_listener_test.rb:
```diff
@@ -0,0 +1,95 @@
+
+      it 'successfully invokes job when passed through request body' do
+        # Stub execute call to avoid invoking Active Job callbacks
+        allow(ActiveJob::Base).to receive(:execute).and_return(nil)
```
can change this from allow to expect since this is specifically what we are testing for here.

MrWellington replied:
Makes sense - first time working with RSpec and still getting a sense of its capabilities. Will need to read up on their docs more.

alextwoods commented on test/aws/rails/middleware/elastic_beanstalk_worker_listener_test.rb:
```diff
@@ -0,0 +1,95 @@
+
+        expect(response[0]).to eq(200)
+        expect(response[1]['Content-Type']).to eq('text/plain')
+        expect(response[2]).to eq(['Successfully ran periodic task ElasticBeanstalkPeriodicTask.'])
```
can we add an expectation that the `perform_now` method gets called?  Something like `expect_any_instance_of(ElasticBeanstalkPeriodicTask).to receive(:perform_now)`

MrWellington replied:
Didn't realize you could do that! Naturally that's what I wanted to validate, I'll add the assertion.

alextwoods commented on test/aws/rails/railtie_test.rb:
```diff
@@ -47,6 +47,55 @@ module Rails
 
+      describe '.add_sqsd_listener_middleware' do
+        it 'adds middleware when AWS_PROCESS_BEANSTALK_WORKER_REQUESTS is set to true' do
+          ENV['AWS_PROCESS_BEANSTALK_WORKER_REQUESTS'] = 'True'
```
Setting the ENV in a test could cause side effects, so if possible would be better to set in a before and unset in an after.  Probably wouldn't ever be an issue here since you're deleting it later in the test - but if the test fails/raises an exception before its deleted it will remain set.

mullermp replied:
An RSpec `after` hook might work

MrWellington replied:
Good point on the exception - using an RSpec hook makes more sense.

alextwoods reviewed (commented):
Thanks for opening this - looks great: well structured, well tested and even well documented!  Just a few minor comments.

MrWellington commented on README.md:
````diff
@@ -355,6 +355,31 @@ require_relative 'config/environment' # load rails
+SQS Daemon provided with each worker instance. The middleware will forward each request and parameters to their appropriate jobs.
+
+To protect against forgeries, these requests will only be processed under the following circumstances:
+1. Environment variable ```AWS_PROCESS_BEANSTALK_WORKER_REQUESTS``` is present and set to true (this should only be set in the worker environment).
````
Technically yes - my intent here was to highlight that the middleware can and should be omitted from web environments open to requests from outside. I'll do some rephrasing here.

MrWellington commented:
Just pushed in some changes addressing the comments above as well as a few stylistic updates recommended by a colleague. 

Only thing to note is I opted for a different change than the JobRunner approach - that class' constructor appears to expect a SQS message object. Given we're not actually polling SQS directly here and the daemon doesn't send all the parameters required to instantiate a message, didn't think it made sense to use here.

Alternatively could consider refactoring JobRunner and its consumers to just supply/take the job's JSON string as it doesn't currently require other fields in that object. In my case wanted to avoid impacting the existing code base as much as possible though.

alextwoods reviewed (approved):
This looks good!  Last thing to do is just add a CHANGELOG.md entry (just add a `* Feature - ` entry under `Unreleased Changes`) and I can then merge this and release as version 3.6.0 of the gem.

MrWellington commented:
Just pushed in the update. Thanks everyone - appreciate the responsiveness and collaboration!

alextwoods commented:
Thanks again for the contribution!  I've merged the changes and they have been released as [aws-sdk-rails 3.6.0](https://rubygems.org/gems/aws-sdk-rails/versions/3.6.0)!
